src/core/modules/impl/ContractHandlerApi.ts

```diff
@@ -47,6 +45,7 @@ export class ContractHandlerApi<State> implements HandlerApi<State> {
 
       const stateCopy = deepCopy(currentResult.state, executionContext.evaluationOptions.useFastCopy);
       this.swGlobal._activeTx = interactionTx;
+      this.swGlobal.caller = interaction.caller; // either contract tx id (for internal writes) or transaction.owner
```

Comment:
worst case we will restore the action.caller - though I prefer to use it this way.